### PR TITLE
build: update snap build

### LINKED ISSUE #931: Update snap script for go v1.16
Please refer to the change in 
https://github.com/edgexfoundry/device-modbus-go/pull/240/files

### PR DESCRIPTION
**Pull request by hahattan**
- update snap build to use go 1.16 via a build-snap
- update snap build to be consistent with other device services

Signed-off-by: Chris Hung <chris@iotechsys.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/device-sdk-c/blob/master/.github/CONTRIBUTING.md

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## New Imports

- [ ] Yes
- [x] No

### CODE REVIEW

File: snap/snapcraft.yaml [hunk 1]
```diff
@@ -131,3 +103,7 @@ parts:
          "$SNAPCRAFT_PART_INSTALL/usr/share/doc/device-simple/Attribution.txt"
       install -DT "./LICENSE" \
          "$SNAPCRAFT_PART_INSTALL/usr/share/doc/device-simple/LICENSE"
+
+  config-common:
+    plugin: dump
+    source: snap/local/runtime-helpers
```
Comment (cloudxxx8) [inline snap/snapcraft.yaml:109]: please add an empty line

**Review by cloudxxx8 — APPROVED** [review]
LGTM

File: Makefile [hunk 2]
```diff
@@ -15,7 +15,6 @@ GOTESTFLAGS?=-race
 GIT_SHA=$(shell git rev-parse HEAD)
 
 build: $(MICROSERVICES)
-	$(GOCGO) install -tags=safe
```
Comment (tonyespy) [inline Makefile:18]: Why was this line removed? It doesn't seem related to the snap build...

File: snap/snapcraft.yaml [hunk 3]
```diff
@@ -25,22 +24,21 @@ architectures:
...
     # would need adapter: full
     adapter: none
-    command: bin/device-simple -confdir $SNAP_DATA/config/device-simple -profile res --registry $CONSUL_ADDR
+    command: bin/device-simple $CONFIG_PRO_ARG $CONF_ARG $REGISTRY_ARG
     # Since we can't include the colon ':' characters in the command above,
     # we need to specify the consul address in an env var like so
     # see https://bugs.launchpad.net/snapd/+bug/1827392 for more details
     environment:
-      CONSUL_ADDR: "consul://localhost:8500"
+      REGISTRY_ARG: "--registry"
```
Comment (tonyespy) [inline snap/snapcraft.yaml:39]: nit: this could just be specified directly on the command line. The only reason we use environment vars for some of the command line args is that they include special chars (eg. ':', '/', '=', ...).

File: snap/snapcraft.yaml [hunk 4]
```diff
@@ -49,52 +47,25 @@ apps:
...
-    prime:
-      - "-*"
+    override-pull: |
+      cd $SNAPCRAFT_PROJECT_DIR
+      if [ -f VERSION ]; then
+        PROJECT_VERSION=$(cat VERSION)
+      else
+        PROJECT_VERSION="0.0.0"
+      fi
+      snapcraftctl set-version ${PROJECT_VERSION}
```
Comment (tonyespy) [inline snap/snapcraft.yaml:60]: This whole section can be simplified as:
```

  version:
    plugin: nil
    source: snap/local
    override-pull: |
      cd $SNAPCRAFT_PROJECT_DIR
      GIT_VERSION=$(git describe --tags --abbrev=0 | sed 's/v//')
      snapcraftctl set-version ${GIT_VERSION}
```

The VERSION file method is really only used for versioning docker containers.

File: snap/snapcraft.yaml [hunk 5]
```diff
@@ -25,22 +24,21 @@ architectures:
...
-    command: bin/device-simple -confdir $SNAP_DATA/config/device-simple -profile res --registry $CONSUL_ADDR
+    command: bin/device-simple $CONFIG_PRO_ARG $CONF_ARG $REGISTRY_ARG
     # Since we can't include the colon ':' characters in the command above,
     # we need to specify the consul address in an env var like so
     # see https://bugs.launchpad.net/snapd/+bug/1827392 for more details
     environment:
-      CONSUL_ADDR: "consul://localhost:8500"
+      REGISTRY_ARG: "--registry"
+      CONFIG_PRO_ARG: "--cp=consul://localhost:8500"
+      CONF_ARG: "--confdir=$SNAP_DATA/config/device-rest-go/res"
```
Comment (tonyespy) [inline snap/snapcraft.yaml:41]: This dir here needs to be `device-simple`, not `device-rest-go`.

File: snap/snapcraft.yaml [hunk 6]
```diff
@@ -106,21 +77,22 @@ parts:
...
       # "ProfilesDir" in combination with the confdir and profile command-line
       # options are sufficient to not need to cd before exec'ing the binary.
       # Change "ProfilesDir" to be under $SNAP_DATA which is writable and 
       # where the install hook copies the config files and device profiles to
       # when the snap is first installed 
       cat "./example/cmd/device-simple/res/configuration.toml" | \
         sed -e s:\"./device-simple.log\":\'\$SNAP_COMMON/device-simple.log\': \
-          -e s:'ProfilesDir = \"./res\"':'ProfilesDir = \"\$SNAP_DATA/config/device-simple/res\"': > \
+        -e s:"ProfilesDir = './res/profiles'":"ProfilesDir = '\$SNAP_DATA/config/device-simple/res/profiles'":  \
+        -e s:"DevicesDir = './res/devices'":"DevicesDir = '\$SNAP_DATA/config/device-simple/res/devices'": > \
```
Comment (tonyespy) [inline snap/snapcraft.yaml:88]: This `sed` statement can just be replaced by a cp command, and then these both can just be overridden with environment variables in the app stanza above. Ex.

    command: bin/device-simple $CONFIG_PRO_ARG $CONF_ARG $REGISTRY_ARG
    # Since we can't include the colon ':' characters in the command above,
    # we need to specify the consul address in an env var like so
    # see https://bugs.launchpad.net/snapd/+bug/1827392 for more details
    environment:
      REGISTRY_ARG: "--registry"
      CONFIG_PRO_ARG: "--cp=consul://localhost:8500"
      CONF_ARG: "--confdir=$SNAP_DATA/config/device-simple/res"
      DEVICE_PROFILESDIR: "$SNAP_DATA/config/device-simple/res/profiles"
      DEVICE_DEVICESDIR: "$SNAP_DATA/config/device-simple/res/devices"

File: snap/snapcraft.yaml [hunk 7]
```diff
@@ -131,3 +103,7 @@ parts:
          "$SNAPCRAFT_PART_INSTALL/usr/share/doc/device-simple/Attribution.txt"
       install -DT "./LICENSE" \
          "$SNAPCRAFT_PART_INSTALL/usr/share/doc/device-simple/LICENSE"
+
+  config-common:
```
Comment (tonyespy) [inline snap/snapcraft.yaml:107]: This stanza can be removed as there aren't any runtime-helpers included in this snap.

**Review by tonyespy — CHANGES_REQUESTED** [review]
Thanks for submitting this PR, most of the changes look good, although I've requested a few changes, so please see my inline comments.

File: Makefile (hunk 2 above)
Comment (hahattan) [reply Makefile]: This change is to be consistent with other device service.
Reverted the change to make this PR only snap-related.

File: snap/snapcraft.yaml [hunk 8]
```diff
@@ -106,21 +74,19 @@ parts:
...
         "$SNAPCRAFT_PART_INSTALL/config/device-simple/res/configuration.toml"
 
-      # Install the example Simpler.Driver device profile
-      cp "./example/cmd/device-simple/res/Simple-Driver.yaml" \
-        "$SNAPCRAFT_PART_INSTALL/config/device-simple/res/Simple-Driver.yaml"
+      # Install the example device and device profile
+      install -DT "./example/cmd/device-simple/res/devices/simple-device.toml" \
+        "$SNAPCRAFT_PART_INSTALL/config/device-simple/res/devices/simple-device.toml"
+      install -DT "./example/cmd/device-simple/res/profiles/Simple-Driver.yaml" \
+        "$SNAPCRAFT_PART_INSTALL/config/device-simple/res/profiles/Simple-Driver.yaml"
```
Comment (tonyespy) [inline snap/snapcraft.yaml:89]: One more change is required to make the snap build properly:

```
$ git diff
diff --git a/snap/snapcraft.yaml b/snap/snapcraft.yaml
index 5042ce7..f2e69ac 100644
--- a/snap/snapcraft.yaml
+++ b/snap/snapcraft.yaml
@@ -93,7 +93,7 @@ parts:
       # licenses in the snap.
       # Note also if the license is changed that the license key in the
       # metadata for the snap at the top of this file should be updated too.
-      install -DT "./Attribution.txt" \
+      install -DT "./example/cmd/device-simple/Attribution.txt" \
          "$SNAPCRAFT_PART_INSTALL/usr/share/doc/device-simple/Attribution.txt"

```

**Review by tonyespy — CHANGES_REQUESTED** [review]
Thanks for the changes! Just one small change is still required (see inline comment).

**Review by tonyespy — APPROVED** [review]
LGTM
